### DISCUSSION TITLE
Groovy Language Server integration

### DISCUSSION THREAD

**Discussion by pennywort** [TypeFox/monaco-languageclient#429] [Q&A]
I'm trying to connect [this Groovy LSP](https://github.com/GroovyLanguageServer/groovy-language-server) with Monaco Editor and not sure that it's possible. 
Can anyone explain conceptual model which represents instances and interaction between them or share some example where Monaco Editor is connected to any LSP implemented on Java?

**Comment by CGNonofr** [comment]
I can affirm it can work as we are using it as well.

You question is very vague, may I suggest you read some documentation about LSP beforehand.

You can probably pretty easily start from the ext example here and replace the json LSP by the one from groovy

**Reply by pennywort** [reply]
I'm very sorry for this.
It's clear to me how to connect typescript based LSP since it works in browser.

I've read about LSP and theoretically understand how it works, it seems I need an example of how (and where?) to run my Java LSP and to respond to editor's requests via JSON-RPC proxy.

**Reply by CGNonofr** [reply]
> It's clear to me how to connect typescript based LSP since it works in browser.

I'm not sure what you are talking about, are you talking about the monaco typescript worker? it doesn't use LSP

> it seems I need an example of how (and where?) to run my Java LSP and to respond to editor's requests via JSON-RPC proxy.

Java or groovy? As I said, there is an example in this repo you can start with

**Reply by kaisalmen** [reply]
Hi @pennywort this issue could help you solve your problem: https://github.com/TypeFox/monaco-languageclient/issues/299

**Reply by kaisalmen** [reply]
... but in issue 299 they solved Java LSP connection problems. @CGNonofr is right you are referring to Groovy LSP above.

**Reply by pennywort** [reply]
> > It's clear to me how to connect typescript based LSP since it works in browser.
> 
> I'm not sure what you are talking about, are you talking about the monaco typescript worker? it doesn't use LSP
> 
> > it seems I need an example of how (and where?) to run my Java LSP and to respond to editor's requests via JSON-RPC proxy.
> 
> Java or groovy? As I said, there is an example in this repo you can start with

Groovy Language Server which implemented as [Java application](https://github.com/GroovyLanguageServer/groovy-language-server). I'm sure it's Java, not Groovy code in this repo

[JSON server](https://github.com/microsoft/vscode/tree/main/extensions/json-language-features/server) is implemented on Typescript

**Reply by CGNonofr** [reply]
The implementation language of the language server is not really important

Did you have a look at the example? it just run a command, you can replace it by any command you want that start a language server

**Reply by pennywort** [reply]
> Hi @pennywort this issue could help you solve your problem: #299

Thank you! Possibly this is what I want, still would see the code:D

**Reply by pennywort** [reply]
> The implementation language of the language server is not really important
> 
> Did you have a look at the example? it just run a command, you can replace it by any command you want that start a language server

Yes, I ran all these examples, it works. But I cannot understand why the language is not important. I'm sure I must run this Java app somewhere and this place would not be a browser

**Reply by CGNonofr** [reply]
I never said it was in the browser, I'm talking about https://github.com/TypeFox/monaco-languageclient/blob/main/packages/examples/server/src/json-server-launcher.ts which start a node process which can be replaced by your java process

**Reply by pennywort** [reply]
I have tried to run Groovy LSP using the command from README.
It's silent in terminal, port is not specified and there's no way to specify, so I cannot call it from client, seems like I need to make some changes in code of this language server.
@kaisalmen @CGNonofr thank you, I'm leaving to think about all this!

**Reply by CGNonofr** [reply]
Most of the language servers don't open a port but communicate through stdin/stdout. Even if it does, it would be a raw socket and not a websocket

It silent in terminal because it's waiting for the initialize message from the client
